OpenRA.Mods.Common/Traits/Health.cs

```diff
@@ -68,43 +69,63 @@ public class Health : IHealth, ISync, ITick, INotifyCreated, INotifyOwnerChanged
 		public Health(ActorInitializer init, HealthInfo info)
 		{
 			Info = info;
-			MaxHP = hp = info.HP > 0 ? info.HP : 1;
+			HP = MaxHP = info.HP > 0 ? info.HP : 1;
 
 			// Cast to long to avoid overflow when multiplying by the health
 			var healthInit = init.GetOrDefault<HealthInit>();
 			if (healthInit != null)
-				hp = (int)(healthInit.Value * (long)MaxHP / 100);
+				HP = (int)(healthInit.Value * (long)MaxHP / 100);
 
 			DisplayHP = hp;
 		}
 
-		public int HP => hp;
+		public int HP
+		{
+			get => hp;
+			private set
+			{
+				hp = value;
+
+				// PERF: DamageState is queried more often than that hp is modified.
+				DamageState = HitPointsToDamangeState(hp, MaxHP);
+				DisplayHP = HitPointsToDisplayHP(hp, DisplayHP);
+			}
+		}
+
 		public int MaxHP { get; }
 
 		public bool IsDead => hp <= 0;
 		public bool RemoveOnDeath = true;
 
-		public DamageState DamageState
+		public DamageState DamageState { get; private set; }
+
+		static DamageState HitPointsToDamangeState(int hp, int maxHP)
 		{
-			get
-			{
-				if (hp == MaxHP)
-					return DamageState.Undamaged;
+			if (hp == maxHP)
+				return DamageState.Undamaged;
 
-				if (hp <= 0)
-					return DamageState.Dead;
+			if (hp <= 0)
+				return DamageState.Dead;
 
-				if (hp * 100L < MaxHP * 25L)
-					return DamageState.Critical;
+			hp *= 100;
+			if (hp < maxHP * 25)
```

Comment:
IIrc this was necessary to avoid overflows with large HP values. Why do you change it back? Can we change it back?

Reply:
See #14644 for context.

Reply:
> The purpose of `displayHP` and `ITick` was to smoothly animate the health bar when actors take damage. This is now broken!

I see. I thought that over time the scale for UI had changed - for some reason. That make sense.

Let me close this then.